### ISSUE TITLE
contradicting information about possible locations for marked content operators

### ISSUE THREAD

**Issue by seehuhn** [pdf-association/pdf-issues#335]
Section 14.6.1, after note 3, states:

> Marked-content operators may appear only between graphics objects in the content stream. They may not occur within a graphics object or between a graphics state operator and its operands.

While this seems clear, there are many examples in the spec, where marked context operators are used within text objects.  For example, the text after table 352 explicitly states that `BT BMC ... EMC ET` is valid, and section 14.9.5 (Expansion of abbreviations and acronyms) has an example where two `BDC ... EMC` pairs are used within the same text object.  (More similar uses within `BT ... ET` are easily found by searching for `BDC`.)

According to the list in section 8.2 (Graphics objects), text objects are one of the five types of PDF graphics objects and according to section 9.4.1, a "text object begins with the BT operator and ends with the ET operator".  Thus, the use of marked-content operators within `BT ... ET` pairs seems to contradict the statement that marked-content operators may appear only between graphics objects.

My guess is that marked-content operators within text objects are in fact allowed, and section 14.6.1 should be clarified to reflect this.

**Comment by petervwyatt** [comment]
Figure 9 does clarify things. 

It's probably best/easiest to explicitly state where marked content operators cannot go: in between operators in path, clipping path, and inline image objects as per Figure 9.

**Comment by ChristianHaider** [comment]
To me, the term "Object" is used in too many informal ways for several things in the spec. 
For me "Content objects" are constituted by a painting operator and its (relevant) graphic state at the time, because only painting operators do change anything "real" on a page.
In this sense, a BT/ET text object is not a real object, but a container for many.
Therefore, the natural place for the marks is right after a painting operator (and at the beginning of the content stream), while respecting the proper nesting in q/Q and BT/ET contexts.

**Comment by petervwyatt** [comment]
I agree that the word "object" is heavily overloaded/over-used in the PDF spec and can be very confusing. `BX`/`EX` for example use the term "section" rather than object, so maybe this might be an acceptable alternative for `BT`/`ET`. However, for the purposes of errata corrections, patching to correct such terminology would be too big a change and too far-reaching - it will need to wait for a full revision of the spec. Sorry.

Once the errata for the OP is done (or not - depending on the PDF TWG), I will either mark this issue as "Parked"  or create a new issue for clarifying the term "object" longer term. Thanks for the comment!

**Comment by petervwyatt** [comment]
PDF TWG want to see precise wording

**Comment by stechio** [comment]
@ChristianHaider:

> For me "Content objects" are constituted by a painting operator and its (relevant) graphic state at the time, because only painting operators do change anything "real" on a page. In this sense, a BT/ET text object is not a real object, but a container for many. 

Your point of view is interesting, I agree with it (my own processor has 2 distinct representations for content streams: *imperative* (corresponding to the procedural, operation-based model described by ISO PDF — here text objects are modeled as-is) and *declarative* (aggregating to "real" objects the relevant graphics state at the time of rendering — here text strings within a text object are modeled as independent objects along with their properties derived from the graphics state)).

@petervwyatt:

> BX/EX for example use the term "section" rather than object, so maybe this might be an acceptable alternative for BT/ET.

I agree that text objects would be more appropriately denominated "text sections".

**Comment by petervwyatt** [comment]
Original problematic paragraph from 14.6.1:

> Marked-content operators may appear only between graphics objects in the content stream. They may not occur within a graphics object or between a graphics state operator and its operands. Marked-content sequences may be nested one within another, but each sequence shall be entirely contained within a single content stream. "Table 352 — Marked-content operators" summarises the marked-content operators.

Proposed change:
> Marked-content operators may appear only at the content stream level or within a text object as shown in "Figure 9 — Graphics objects". ~~They may not occur between an operator and its operands.~~ Marked-content sequences may be nested one within another, but each sequence shall be entirely contained within a single content stream. "Table 352 — Marked-content operators" summarises the marked-content operators.

Note that the 2nd sentence (shown as ~~strikethrough~~) now reduces to something that I think is meaningless and is effectively stated elsewhere (there is no persistent stack in PDF), so I propose to delete it also. But left in-situ in case others feel differently.

Note that the explanation on "nicely" nesting things is at the end of this subclause and I don't want to do larger surgery to move slabs of text around for an errata. As noted above, this is just a point-fix for the original problem and we'll track a larger change for when the spec is open again for significant edits.

**Comment by petervwyatt** [comment]
PDF TWG agree
